## MNIST Classification 

#### Grid Search for Best Hyperparameters

In [13]:
from sklearn.datasets import load_digits
from sklearn.model_selection import GridSearchCV
from sklearn import svm

In [32]:
import multiprocessing

In [33]:
cpus = multiprocessing.cpu_count()
print(cpus)

4


In [23]:
params = [
    {
        'kernel':['linear','rbf','poly','sigmoid'],
        'C':[0.1,0.2,0.5,1.0,2.0,5.0]
    }   
]


# Dataset Prepare

In [25]:
data = load_digits()
X = data.data
Y  = data.target
print(X.shape,Y.shape)

(1797, 64) (1797,)


## Logistic Regression

In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [28]:
lr = LogisticRegression()
lr.fit(X,Y)
cross_val_score(lr,X,Y,scoring="accuracy",cv=5).mean()

0.9210188113360701

### Logistic Regression gives around 92% accuracy

In [29]:
from sklearn import svm

In [31]:
svc = svm.SVC()
svc.fit(X,Y)
cross_val_score(svc,X,Y,scoring="accuracy",cv=5).mean()

0.44878680061604637

### Pretty Bad Result :(?

## Right Set of Parameters !

In [34]:
gs = GridSearchCV(estimator=svm.SVC(),param_grid=params,scoring="accuracy",cv=5,n_jobs = cpus)

In [35]:
gs.fit(X,Y)

GridSearchCV(cv=5, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=4,
       param_grid=[{'kernel': ['linear', 'rbf', 'poly', 'sigmoid'], 'C': [0.1, 0.2, 0.5, 1.0, 2.0, 5.0]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=0)

In [36]:
gs.best_estimator_

SVC(C=0.1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [37]:
gs.best_score_

0.9699499165275459

## SVM with Polynomial Kernel, C=0.1 gives 97% accuracy on the Dataset!